UnitySDK/Assets/ML-Agents/Scripts/LearningBrain.cs

```diff
@@ -19,468 +12,138 @@ namespace MLAgents
-                    var m = new float[memorySize];
-                    for (int j = 0; j < memorySize; j++)
-                    {
-                        m[j] = recurrentTensor[i, j];
-                    }
-
-                    agent.UpdateMemoriesAction(m.ToList());
-                    i++;
-                }
-            }
-
-            
-            if (hasValueEstimate)
-            {
-                float[,] value_estimates = new float[currentBatchSize,1];
-                if (hasRecurrent)
-                {
-                    value_estimates = networkOutput[2].GetValue() as float[,];
-                }
-                else
-                {
-                    value_estimates = networkOutput[1].GetValue() as float[,];
+                    throw new UnityAgentsException(
+                        "Unknow tensor expected as output : "+tensor.Name);
                 }
                 
-                var i = 0;
-                foreach (Agent agent in agentList)
-                {
-                    agent.UpdateValueAction(value_estimates[i,0]);
-                }
+                _tensorGeneratorInvoker[tensor.Name].Execute(tensor, currentBatchSize, agentInfos);
             }
 
-            if (brainParameters.vectorActionSpaceType == SpaceType.continuous)
-            {
-                var output = networkOutput[0].GetValue() as float[,];
-                var i = 0;
-                foreach (Agent agent in agentList)
-                {
-                    var a = new float[brainParameters.vectorActionSize[0]];
-                    for (int j = 0; j < brainParameters.vectorActionSize[0]; j++)
-                    {
-                        a[j] = output[i, j];
-                    }
+            // Execute the Model
+            _engine.ExecuteGraph(_inferenceInputs, _inferenceOutputs);
 
-                    agent.UpdateVectorAction(a);
-                    i++;
-                }
-            }
-            else if (brainParameters.vectorActionSpaceType == SpaceType.discrete)
+            // Update the outputs
+            foreach (var tensor in _inferenceOutputs)
             {
-                long[,] output = networkOutput[0].GetValue() as long[,];
-                var i = 0;
-                foreach (Agent agent in agentList)
+                if (!_outputTensorApplierInvoker.ContainsKey(tensor.Name))
                 {
-                    var actSize = brainParameters.vectorActionSize.Length;
-                    var a = new float[actSize];
-                    for (int actIdx = 0; actIdx < actSize; actIdx++)
-                    {
-                        a[actIdx] = output[i, actIdx];
-                    }
-                    agent.UpdateVectorAction(a);
-                    i++;
+                    throw new UnityAgentsException(
+                        "Unknow tensor expected as output : "+tensor.Name);
                 }
+                _outputTensorApplierInvoker[tensor.Name].Execute(tensor, agentInfos);
```

Comment:
Might make sense to get rid of the Invoker here as well (even though it's less important since you don't have a copy paste loop like in the generator)